File: datasets/synthetic/PT-PY-REPO-002/src/services/file_service.py
import os

from utils.file_utils import read_file

BASE_DIR = "/var/www/files"


class FileService:
    def _resolve_safe_path(self, filename: str) -> str:
        clean = os.path.normpath(filename)
        if os.path.isabs(clean):
            raise ValueError("absolute path rejected")
        joined = os.path.join(BASE_DIR, clean)
        resolved = os.path.realpath(joined)
        real_base = os.path.realpath(BASE_DIR)
        if not resolved.startswith(real_base + os.sep):
            raise ValueError("path escape detected")
        return resolved

    def get_file_content(self, filename: str) -> bytes | None:
        try:
            safe_path = self._resolve_safe_path(filename)
        except ValueError:
            return None
        return read_file(safe_path)

File: datasets/synthetic/PT-JA-REPO-002/src/main/java/com/example/demo/service/FileService.java
package com.example.demo.service;

import com.example.demo.util.FileUtil;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;

public class FileService {

    private static final Path BASE_DIR =
            Paths.get("/var/www/files").toAbsolutePath();

    private Path resolveSafePath(String filename) throws IOException {
        Path resolved = BASE_DIR.resolve(filename).normalize();
        if (resolved.isAbsolute() && !resolved.startsWith(BASE_DIR)) {
            throw new SecurityException("path escape detected");
        }
        Path real = resolved.toRealPath();
        Path realBase = BASE_DIR.toRealPath();
        if (!real.startsWith(realBase)) {
            throw new SecurityException("path escape detected after symlink resolution");
        }
        return real;
    }

    public byte[] getFileContent(String filename) {
        Path safePath;
        try {
            safePath = resolveSafePath(filename);
        } catch (SecurityException | IOException e) {
            return null;
        }
        return FileUtil.readFile(safePath);
    }
}
